File: api/src/schemas/transaction.py
import datetime
import uuid
from decimal import Decimal
from typing import Any

from pydantic import BaseModel, ConfigDict, field_validator

from .allocation import AllocationResponse


class TransactionCreate(BaseModel):
    date: datetime.date
    description: str
    total_amount: Decimal
    currency_code: str = "USD"
    source: str = "manual"
    external_id: str | None = None
    receipt_file_path: str | None = None
    is_approved: bool = False

    @field_validator("external_id", mode="before")
    @classmethod
    def empty_str_to_none(cls, v: Any) -> Any:
        if isinstance(v, str) and not v.strip():
            return None
        return v


class TransactionUpdate(BaseModel):
    date: datetime.date | None = None
    description: str | None = None
    total_amount: Decimal | None = None
    currency_code: str | None = None
    source: str | None = None
    external_id: str | None = None
    receipt_file_path: str | None = None
    is_approved: bool | None = None

    @field_validator("external_id", mode="before")
    @classmethod
    def empty_str_to_none(cls, v: Any) -> Any:
        if isinstance(v, str) and not v.strip():
            return None
        return v


class TransactionApproveRequest(BaseModel):
    is_approved: bool | None = None


class TransactionResponse(BaseModel):
    id: uuid.UUID
    source: str
    external_id: str | None = None
    date: datetime.date
    description: str
    total_amount: Decimal
    currency_code: str
    receipt_file_path: str | None = None
    is_approved: bool
    allocations: list[AllocationResponse] = []

    model_config = ConfigDict(from_attributes=True)


class PaginatedTransactionsResponse(BaseModel):
    items: list[TransactionResponse]
    total: int
    page: int
    page_size: int
    total_pages: int

File: web/src/types/transaction.ts
export type SyncStatus = "PENDING" | "SYNCED"

export interface Allocation {
  id: string
  transaction_id: string
  amount: string | number
  is_personal: boolean
  company_id: string | null
  sync_status: SyncStatus
}

export interface Transaction {
  id: string
  source: string
  external_id: string | null
  date: string
  description: string
  total_amount: string | number
  currency_code: string
  receipt_file_path: string | null
  is_approved: boolean
  allocations: Allocation[]
}

export interface PaginatedTransactionsResponse {
  items: Transaction[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

export interface TransactionFilterParams {
  source?: string
  is_approved?: boolean | null
  start_date?: string
  end_date?: string
  company_id?: string | null
}

export interface TransactionQueryParams extends TransactionFilterParams {
  page?: number
  page_size?: number
  rowsPerPage?: number
  sort_by?: string
  sortBy?: string
  descending?: boolean
  order?: "asc" | "desc"
}
